user/oprofile/doc/ophelp.1.in
.TH OPHELP 1 "@DATE@" "oprofile @VERSION@"
.UC 4
.SH NAME
ophelp \- list OProfile events
.SH SYNOPSIS
.br
.B ophelp
[
.I options
]
[event name]
.SH DESCRIPTION

By default,
.B ophelp
lists the available performance counter options. If you
give it a symbolic event name, it will return the hardware
value (e.g. "ophelp DATA_MEM_REFS").

.SH OPTIONS
.TP
.BI "--cpu-type / -c"
Show the events for the given numerical CPU type.
.br
.TP
.BI "--get-cpu-type / -r"
Show the symbolic CPU name.
.br
.TP
.BI "--help / -? / --usage"
Show help message.
.br
.TP
.BI "--version / -v"
Show version.

.SH ENVIRONMENT
No special environment variables are recognised by ophelp.

.SH FILES
.TP
.I $prefix/share/oprofile/
Event description files used by OProfile.
.TP
.I /var/lib/oprofile/samples/
The location of the generated sample files.

.SH VERSION
.TP
This man page is current for @PACKAGE@-@VERSION@.

.SH SEE ALSO
.BR @OP_DOCDIR@,
.BR oprofile(1)


freeswan/CHANGES
New and notable in 1.97:

Sanity check added for at least one of AH or ESP more for checking if
"make *config" has even been run.

Fixed 2.2 local IKE fragmentation blackhole.  Still won't work if
iptraf or another pcap app is running.

MCR added ikeping(8).

Applied DHR's tunnel patch to streamline IKE/specialSA processing.

Added local and remote IKE bypass tests to the UML test suite.

UML KLIPS %trap* tests now check for upbound pf_key messages.

An ICMP ICMP_DEST_UNREACH, ICMP_PKT_FILTERED is now sent upon %reject.

Added a UML input file with first and last packets being outside the 
covered range to "sandwich" the packets under test.

The default _updown script has been cleaned up slightly, reducing
internal duplication that arose accidentally.

When doing --add on an RSA-signature conn, ipsec_auto now tells Pluto
about the keys before, not after, telling it about the conn.  The main
consequence of this is that if something is wrong with one of the keys,
the conn itself does *not* get loaded.  This should reduce confusion.

The rp_filter checks during KLIPS startup have been fixed to be a bit less
paranoid; rp_filter is thought to cause problems only on the physical
interfaces, not on the ipsecN interfaces. 

The configuration process has been fixed to be a bit more paranoid.  Some
users have had mysterious compile failures with complaints involving names
like MD5_CTX; they seem to have resulted from kernel configuration changes
somehow getting into .config but not into include/linux/autoconf.h (which
is what actually controls most of the kernel compile).  The post-config
checking now verifies not only that .config has the right stuff in it, but
that autoconf.h does too.  (We still don't see *how* this discrepancy can
arise, but this will at least catch it when it does.)

The details of linking the KLIPS source into the kernel are now under
control of a configuration variable in Makefile.inc, to make it easier
for distribution builders to change.

The top-level Makefile now includes cleverness which assigns a date-based
version code if a build is done from a raw CVS checkout.

The library now includes a new function, ttodatav(), like ttodata() but
in some cases it gives more verbose error reports.  Also, there are now
a couple of functions which distill an RSA public key to a "key ID" for
use in messages, a crypto-quality pseudo-random-number generator, and a
function that converts a single address into a (singleton) subnet.

Barf's secret/key-censoring utilities have been tidied up so that what
they emit matches the key-ID library functions. 

The ipsec command now supplies a couple more environment variables to
things it calls, and this has reduced the number of things which need to
be edited at build time to know about configuration variables. 

The configuration reader now allows parameter names to begin with "_";
these are strictly reserved for internal development purposes and will
not in general be documented.


As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.96:

Compatibility with the 2.0.xx kernels (under Red Hat 5.2) is broken.
This was accidental, and we have some hopes of fixing it, although it
won't last forever even if we do.  See README.

Our "IKE Implementation Issues" IETF draft has been added to our docs.

Newhostkey now has a mandatory --output "option" specifying a filename
(although the magic filename "-" means "standard output"), and sets
restrictive permissions on newly-created files to try to keep private keys
truly private. 

Pluto now gracefully handles a couple of odd error cases (missing private
keys, network errors at just the wrong time in negotiations) which used to
cause it to die, provoking an automatic restart of the IPsec subsystem. 

Makefile.inc now has more detailed control over where manpages go, to ease
integration into distributions.  The name of the destination directory for
RPMs is also now centralized there.

Rsasigkey now checks for certain obscure internal failures, which can
happen if --random is used to supply a source of random bits that aren't
really very random, and complains accordingly.

Barf's eroute table, formerly sorted by source, is now sorted by destination.
This is probably a bit clearer for hosts with lots of eroutes.

Use of the %reject target in a shunt eroute now, as long planned, actually
causes an ICMP rejection message (Destination Unreachable:  Communication
Administratively Prohibited). 

KLIPS now declares its module license to be "Dual BSD/GPL".  (It is 
dual because the libdes and zlib code is BSD licensed, while KLIPS itself 
is GPL).  Whether this is exactly right is not yet clear, but at least it
stops the latest module loaders from whining.

The UML test stuff continues to grow and add new test cases.

As usual, there are assorted small bug fixes and improvements to docs
and messages.



New and notable in 1.95:

The last remnants of the "%hold" bug, which broke 1.93 and 1.94, have
(we think) been dealt with.

Minor bugs in the RPM-building stuff have been fixed, and it has been
updated to deal with Red Hat's abrupt removal of the --buildroot option
from rpm.

Pluto now supports the ability to delete an instance of a conn, rather
than just the whole conn (for cases, notably OE, where a single connection
description can be involved in multiple connections).  There is no support
for this in ipsec_auto yet.

Pluto's routine (non-debug) logging is now somewhat terser, and some of
the more boring messages are produced only if --debug-lifecycle is turned
on (this is a temporary hack, it's not really debugging output).

Pluto's innards have been reorganized somewhat, in preparation for making
DNS lookups asynchronous.  There should be no user-visible effects.

Pluto now ignores IPv6 link-local interface addresses; we think that they
are never relevant, and trying to bind to them causes difficulties.  This

user/oprofile/utils/ophelp.c
/**
 * @file ophelp.c
 * Print out PMC event information
 *
 * @remark Copyright 2002 OProfile authors
 * @remark Read the file COPYING
 *
 * @author John Levon
 * @author Philippe Elie
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

#include "op_version.h"
#include "op_events.h"
#include "op_popt.h"
#include "op_cpufreq.h"
#include "op_hw_config.h"
#include "op_string.h"
#include "op_alloc_counter.h"
#include "op_parse_event.h"
#include "op_libiberty.h"

static char const ** chosen_events;
struct parsed_event * parsed_events;
static op_cpu cpu_type = CPU_NO_GOOD;
static char * cpu_string;
static int callgraph_depth;

static poptContext optcon;


/// return the Hamming weight (number of set bits)
static size_t hweight(size_t mask)
{
	size_t count = 0;

	while (mask) {
		mask &= mask - 1;
		count++;
	}

	return count;
}


/**
 * help_for_event - output event name and description
 * @param i  event number
 *
 * output an help string for the event @i
 */
static void help_for_event(struct op_event * event)
{
	uint i, j;
	uint mask;

	size_t nr_counters = op_get_nr_counters(cpu_type);

	printf("%s", event->name);

	printf(": (counter: ");

	mask = event->counter_mask;
	if (hweight(mask) == nr_counters) {
		printf("all");
	} else {
		for (i = 0; i < CHAR_BIT * sizeof(event->counter_mask); ++i) {
			if (mask & (1 << i)) {
				printf("%d", i);
				mask &= ~(1 << i);
				if (mask)
					printf(", ");
			}
		}
	}

	printf(")\n\t%s (min count: %d)\n", event->desc, event->min_count);

	if (strcmp(event->unit->name, "zero")) {

		printf("\tUnit masks (default 0x%x)\n",
		       event->unit->default_mask);
		printf("\t----------\n");

		for (j = 0; j < event->unit->num; j++) {
			printf("\t0x%.2x: %s\n",
			       event->unit->um[j].value,
			       event->unit->um[j].desc);
		}
	}
}


static void check_event(struct parsed_event * pev,
			struct op_event const * event)
{
	int ret;
	int min_count;
	int const callgraph_min_count_scale = 15;

	if (!event) {
		fprintf(stderr, "No event named %s is available.\n",
		        pev->name);
		exit(EXIT_FAILURE);
	}

	ret = op_check_events(0, event->val, pev->unit_mask, cpu_type);

	if (ret & OP_INVALID_UM) {
		fprintf(stderr, "Invalid unit mask 0x%x for event %s\n",
		        pev->unit_mask, pev->name);
		exit(EXIT_FAILURE);
	}

	min_count = event->min_count;
	if (callgraph_depth)
		min_count *= callgraph_min_count_scale;
	if (pev->count < min_count) {
		fprintf(stderr, "Count %d for event %s is below the "
		        "minimum %d\n", pev->count, pev->name, min_count);
		exit(EXIT_FAILURE);
	}
}


static void resolve_events(void)
{
	size_t count;
	size_t i, j;
	size_t * counter_map;
	size_t nr_counters = op_get_nr_counters(cpu_type);
	struct op_event const * selected_events[nr_counters];

	count = parse_events(parsed_events, nr_counters, chosen_events);
	if (count > nr_counters) {
		fprintf(stderr, "Not enough hardware counters.\n");
		exit(EXIT_FAILURE);
	}

	for (i = 0; i < count; ++i) {
		for (j = i + 1; j < count; ++j) {
			struct parsed_event * pev1 = &parsed_events[i];
			struct parsed_event * pev2 = &parsed_events[j];

			if (!strcmp(pev1->name, pev2->name) &&
			    pev1->count == pev2->count &&
			    pev1->unit_mask == pev2->unit_mask &&
			    pev1->kernel == pev2->kernel &&
			    pev1->user == pev2->user) {
				fprintf(stderr, "All events must be distinct.\n");
				exit(EXIT_FAILURE);
			}
		}
	}

	for (i = 0; i < count; ++i) {
		struct parsed_event * pev = &parsed_events[i];

		selected_events[i] = find_event_by_name(pev->name);

		check_event(pev, selected_events[i]);
	}

	counter_map = map_event_to_counter(selected_events, count, cpu_type);

	if (!counter_map) {
		fprintf(stderr, "Couldn't allocate hardware counters for the selected events.\n");
		exit(EXIT_FAILURE);
	}

	for (i = 0; i < count; ++i)
		printf("%d ", (unsigned int) counter_map[i]);
	printf("\n");

	free(counter_map);
}


static void show_unit_mask(void)
{
	struct op_event * event;
	size_t count;
	size_t nr_counter = op_get_nr_counters(cpu_type);

	count = parse_events(parsed_events, nr_counter, chosen_events);
	if (count > 1) {
		fprintf(stderr, "More than one event specified.\n");
		exit(EXIT_FAILURE);
	}

	event = find_event_by_name(parsed_events[0].name);

	if (!event) {
		fprintf(stderr, "No such event found.\n");
		exit(EXIT_FAILURE);
	}

	printf("%d\n", event->unit->default_mask);
}


static void show_default_event(void)
{
	struct op_default_event_descr descr;

	op_default_event(cpu_type, &descr);

	if (descr.name[0] == '\0')
		return;

	printf("%s:%lu:%lu:1:1\n", descr.name, descr.count, descr.um);
}


static int show_vers;
static int get_cpu_type;
static int check_events;
static int unit_mask;
static int get_default_event;

static struct poptOption options[] = {
	{ "cpu-type", 'c', POPT_ARG_STRING, &cpu_string, 0,
	  "use the given CPU type", "cpu type", },
	{ "check-events", 'e', POPT_ARG_NONE, &check_events, 0,
	  "check the given event descriptions for validity", NULL, },
	{ "unit-mask", 'u', POPT_ARG_NONE, &unit_mask, 0,
	  "default unit mask for the given event", NULL, },
	{ "get-cpu-type", 'r', POPT_ARG_NONE, &get_cpu_type, 0,
	  "show the auto-detected CPU type", NULL, },
	{ "get-default-event", 'd', POPT_ARG_NONE, &get_default_event, 0,
	  "get the default event", NULL, },
	{ "callgraph", '\0', POPT_ARG_INT, &callgraph_depth, 0,
	  "use this callgraph depth", "callgraph depth", },
	{ "version", 'v', POPT_ARG_NONE, &show_vers, 0,
	   "show version", NULL, },
	POPT_AUTOHELP
	{ NULL, 0, 0, NULL, 0, NULL, NULL, },
};

/**
 * get_options - process command line
 * @param argc  program arg count
 * @param argv  program arg array
 *
 * Process the arguments, fatally complaining on error.
 */
static void get_options(int argc, char const * argv[])
{
	optcon = op_poptGetContext(NULL, argc, argv, options, 0);

	if (show_vers)
		show_version(argv[0]);

	/* non-option, must be a valid event name or event specs */
	chosen_events = poptGetArgs(optcon);

	/* don't free the context now, we need chosen_events */
}


/** make valgrind happy */
static void cleanup(void)
{
	int i;
	for (i = 0; i < op_get_nr_counters(cpu_type); ++i) {
		if (parsed_events[i].name)
			free(parsed_events[i].name);
	}
	op_free_events();
	if (optcon)
		poptFreeContext(optcon);
	if (parsed_events)
		free(parsed_events);
}


int main(int argc, char const * argv[])
{
	struct list_head * events;
	struct list_head * pos;
	char const * pretty;
	size_t nr_counter;

	atexit(cleanup);

	get_options(argc, argv);

	/* usefull for testing purpose to allow to force the cpu type
	 * with --cpu-type */
	if (cpu_string) {
		cpu_type = op_get_cpu_number(cpu_string);
	} else {
		cpu_type = op_get_cpu_type();
	}

	if (cpu_type == CPU_NO_GOOD) {
		fprintf(stderr, "cpu_type '%s' is not valid\n",
		        cpu_string ? cpu_string : "unset");
		exit(EXIT_FAILURE);
	}

	nr_counter = op_get_nr_counters(cpu_type);
	parsed_events = xcalloc(nr_counter, sizeof(struct parsed_event));

	pretty = op_get_cpu_type_str(cpu_type);

	if (get_cpu_type) {
		printf("%s\n", pretty);
		exit(EXIT_SUCCESS);
	}

	if (get_default_event) {
		show_default_event();
		exit(EXIT_SUCCESS);
	}

	if (cpu_type == CPU_TIMER_INT) {
		if (!check_events)
			printf("Using timer interrupt.\n");
		exit(EXIT_SUCCESS);
	}

	events = op_events(cpu_type);

	if (!chosen_events && (unit_mask || check_events)) {
		fprintf(stderr, "No events given.\n");
		exit(EXIT_FAILURE);
	}

	if (unit_mask) {
		show_unit_mask();
		exit(EXIT_SUCCESS);
	}

	if (check_events) {
		resolve_events();
		exit(EXIT_SUCCESS);
	}

	/* without --check-events, the only argument must be an event name */
	if (chosen_events && chosen_events[0]) {
		if (chosen_events[1]) {
			fprintf(stderr, "Too many arguments.\n");
			exit(EXIT_FAILURE);
		}

		list_for_each(pos, events) {
			struct op_event * event = list_entry(pos, struct op_event, event_next);

			if (strcmp(event->name, chosen_events[0]) == 0) {
				char const * map = find_mapping_for_event(event->val, cpu_type);
				if (map) {
					printf("%d %s\n", event->val, map);
				} else {
					printf("%d\n", event->val);
				}
				exit(EXIT_SUCCESS);
			}
		}
		fprintf(stderr, "No such event \"%s\"\n", chosen_events[0]);
		exit(EXIT_FAILURE);
	}

	/* default: list all events */

	printf("oprofile: available events for CPU type \"%s\"\n\n", pretty);
	switch (cpu_type) {
	case CPU_HAMMER:
		break;
	case CPU_ATHLON:
		printf ("See AMD document x86 optimisation guide (22007.pdf), Appendix D\n\n");
		break;
	case CPU_PPRO:
	case CPU_PII:
	case CPU_PIII:
	case CPU_P6_MOBILE:
	case CPU_P4:
	case CPU_P4_HT2:
	case CPU_CORE:
	case CPU_CORE_2:
		printf("See Intel Architecture Developer's Manual Volume 3, Appendix A and\n"
		"Intel Architecture Optimization Reference Manual (730795-001)\n\n");
		break;
	case CPU_IA64:
	case CPU_IA64_1:
	case CPU_IA64_2:
		printf("See Intel Itanium Processor Reference Manual\n"
		       "for Software Development (Document 245320-003),\n"
		       "Intel Itanium Processor Reference Manual\n"
		       "for Software Optimization (Document 245473-003),\n"
		       "Intel Itanium 2 Processor Reference Manual\n"
		       "for Software Development and Optimization (Document 251110-001)\n\n");
		break;
	case CPU_AXP_EV4:
	case CPU_AXP_EV5:
	case CPU_AXP_PCA56:
	case CPU_AXP_EV6:
	case CPU_AXP_EV67:
		printf("See Alpha Architecture Reference Manual\n"
		       "ftp://ftp.compaq.com/pub/products/alphaCPUdocs/alpha_arch_ref.pdf\n");
		break;
	case CPU_ARM_XSCALE1:
	case CPU_ARM_XSCALE2:
		printf("See Intel XScale Core Developer's Manual\n"
		       "Chapter 8 Performance Monitoring\n");
		break;
		break;

	case CPU_PPC64_POWER4:
	case CPU_PPC64_POWER5:
	case CPU_PPC64_POWER5p:
	case CPU_PPC64_970:
		printf("Obtain PowerPC64 processor documentation at:\n"
			"http://www-306.ibm.com/chips/techlib/techlib.nsf/productfamilies/PowerPC\n");
		break;

	case CPU_MIPS_20K:
		printf("See Programming the MIPS64 20Kc Processor Core User's "
		       "manual available from www.mips.com\n");
		break;
	case CPU_MIPS_24K:
		printf("See Programming the MIPS32 24K Core "
		       "available from www.mips.com\n");
		break;
	case CPU_MIPS_25K:
		printf("See Programming the MIPS64 25Kf Processor Core User's "
		       "manual available from www.mips.com\n");
		break;
	case CPU_MIPS_34K:
		printf("See Programming the MIPS32 34K Core Family "
		       "available from www.mips.com\n");
		break;
	case CPU_MIPS_5K:
		printf("See Programming the MIPS64 5K Processor Core Family "
		       "Software User's manual available from www.mips.com\n");
		break;
	case CPU_MIPS_R10000:
	case CPU_MIPS_R12000:
		printf("See NEC R10000 / R12000 User's Manual\n"
		       "http://www.necelam.com/docs/files/U10278EJ3V0UM00.pdf\n");
		break;
	case CPU_MIPS_RM7000:
		printf("See RM7000 Family User Manual "
		       "available from www.pmc-sierra.com\n");
		break;
	case CPU_MIPS_RM9000:
		printf("See RM9000x2 Family User Manual "
		       "available from www.pmc-sierra.com\n");
		break;
	case CPU_MIPS_SB1:
	case CPU_MIPS_VR5432:
		printf("See NEC VR5443 User's Manual, Volume 1\n"
		       "http://www.necelam.com/docs/files/1375_V1.pdf\n");
		break;
	case CPU_MIPS_VR5500:
		printf("See NEC R10000 / R12000 User's Manual\n"
		     "http://www.necel.com/nesdis/image/U16677EJ3V0UM00.pdf\n");
		break;

	case CPU_PPC_E500:
	case CPU_PPC_E500_2:
		printf("See PowerPC e500 Core Complex Reference Manual\n"
			"Chapter 7: Performance Monitor\n"
			"Downloadable from http://www.freescale.com\n");
		break;

	case CPU_PPC_7450:
		printf("See MPC7450 RISC Microprocessor Family User's Manual\n"
		       "Downloadable from http://www.freescale.com\n");
		break;

	case CPU_RTC:
		break;

	// don't use default, if someone add a cpu he wants a compiler warning
	// if he forgets to handle it here.
	case CPU_TIMER_INT:
	case CPU_NO_GOOD:
	case MAX_CPU_TYPE:
		printf("%d is not a valid processor type.\n", cpu_type);
		break;
	}

	list_for_each(pos, events) {
		struct op_event * event = list_entry(pos, struct op_event, event_next);
		help_for_event(event);
	}

	return EXIT_SUCCESS;
}
